core/hotspot/rule_manager.go

```diff
@@ -98,13 +99,79 @@ func onRuleUpdate(rules []*Rule) (err error) {
 		}
 	}()
 
+	newRuleMap := make(map[string][]*Rule)
+	for _, r := range rules {
+		if err := IsValidRule(r); err != nil {
+			logger.Warnf("Ignoring invalid hotspot rule when loading new rules, rule: %s, reason: %s", r.String(), err.Error())
+			continue
+		}
+		res := r.ResourceName()
+		ruleSet, ok := newRuleMap[res]
+		if !ok {
+			ruleSet = make([]*Rule, 0, 1)
+		}
+		ruleSet = append(ruleSet, r)
+		newRuleMap[res] = ruleSet
+	}
+
+	m := make(trafficControllerMap)
+	for res, rules := range newRuleMap {
+		m[res] = make([]TrafficShapingController, 0, len(rules))
+	}
+
+	start := util.CurrentTimeNano()
 	tcMux.Lock()
-	defer tcMux.Unlock()
+	defer func() {
+		tcMux.Unlock()
+		logger.Infof("Updating hotspot rule spends %d ns.", util.CurrentTimeNano()-start)
+		logRuleUpdate(m)
```

Comment:
Here we may need to handle cases regarding panic outside (where the log is not required).